File: c/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int start) {
  if (start <= 0)
    return ERROR_VALUE;

  int ct = 0;
  while (start != 1) {
    ct++;
    if (start % 2 == 0) {
      start /= 2;
    } else {
      start = start * 3 + 1;
    }
  }
  return ct;
}

File: go/collatz_conjecture.go
package collatzconjecture

import "errors"

// CollatzConjecture return the number of steps required to reach 1
func CollatzConjecture(n int) (int, error) {
	if n < 1 {
		return 0, errors.New("error")
	}

	step := 0
	for n != 1 {
		step++

		if n%2 == 0 {
			n /= 2
		} else {
			n = 3*n + 1
		}
	}

	return step, nil
}
